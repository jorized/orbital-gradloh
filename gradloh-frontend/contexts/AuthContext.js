import React, { createContext, useState } from 'react';
import * as SecureStore from 'expo-secure-store';
const AuthContext = createContext(null);
const { Provider } = AuthContext;

const AuthProvider = ({ children }) => {


	const [authState, setAuthState] = useState({
		accessToken: null,
		refreshToken: null,
		authenticated: null
	});

	const logout = async () => {

		await SecureStore.deleteItemAsync('token');
		await SecureStore.deleteItemAsync('userdetails');
		await SecureStore.deleteItemAsync('selectedGradient');
		setAuthState({
			accessToken: null,
			refreshToken: null,
			authenticated: false,
			firstTimeUser: false
		});
	};

	const getAccessToken = () => {
		return authState.accessToken;
	};


	return (
		<Provider
			value={{
				authState,
				getAccessToken,
				setAuthState,
				logout,
			}}
		>
			{children}
		</Provider>
	);
};

export { AuthContext, AuthProvider };
