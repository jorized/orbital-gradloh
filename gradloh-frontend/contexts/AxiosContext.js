import React, { createContext, useContext, useEffect } from 'react';
import axios from 'axios';
import { AuthContext } from './AuthContext';
import { LoadingContext } from './LoadingContext'; // Import the Loading Context
import createAuthRefreshInterceptor from 'axios-auth-refresh';
import * as SecureStore from 'expo-secure-store';

const AxiosContext = createContext();
const authApiUrl = process.env.EXPO_PUBLIC_API_AUTH_URL;
const publicApiUrl = process.env.EXPO_PUBLIC_API_URL;
const { Provider } = AxiosContext;

const AxiosProvider = ({ children }) => {
  const authContext = useContext(AuthContext);
  const { setIsLoading } = useContext(LoadingContext); // Use the Loading Context
  const requestQueue = [];
  let isRefreshing = false;

  const processQueue = (error, token = null) => {
    requestQueue.forEach(promise => {
      if (error) {
        promise.reject(error);
      } else {
        promise.resolve(token);
      }
    });
    requestQueue.length = 0;
  };

      //Those who are not, use this route
    const publicAxios = axios.create({
      baseURL: authApiUrl
    });

  const authAxios = axios.create({
    baseURL: publicApiUrl,
    headers: {
      Authorization: `Bearer ${authContext.getAccessToken()}`,
    },
  });

  authAxios.interceptors.response.use(null, async (error) => {
    const originalRequest = error.config;
    if (error.response.status === 403 && !originalRequest._retry) {
      originalRequest._retry = true;
      if (!isRefreshing) {
        isRefreshing = true;
        setIsLoading(true); // Show loading indicator

        const data = {
          refreshToken: authContext.authState.refreshToken,
        };
        const options = {
          method: 'POST',
          data,
          url: `${authApiUrl}/refresh`,
        };

        try {
          const tokenRefreshResponse = await axios(options);
          isRefreshing = false;
          setIsLoading(false); // Hide loading indicator

          authContext.setAuthState({
            ...authContext.authState,
            accessToken: tokenRefreshResponse.data.accessToken,
          });

          await SecureStore.setItemAsync(
            'token',
            JSON.stringify({
              accessToken: tokenRefreshResponse.data.accessToken,
              refreshToken: authContext.authState.refreshToken,
            })
          );

          processQueue(null, tokenRefreshResponse.data.accessToken);

          originalRequest.headers.Authorization =
            'Bearer ' + tokenRefreshResponse.data.accessToken;

          return axios(originalRequest);
        } catch (e) {
          isRefreshing = false;
          setIsLoading(false); // Hide loading indicator
          processQueue(e, null);
          throw e;
        }
      }

      const retryOriginalRequest = new Promise((resolve, reject) => {
        requestQueue.push({ resolve, reject });
      });

      return retryOriginalRequest
        .then(token => {
          originalRequest.headers.Authorization = 'Bearer ' + token;
          return axios(originalRequest);
        })
        .catch(err => {
          return Promise.reject(err);
        });
    }
    return Promise.reject(error);
  });

  return (
    <Provider value={{ authAxios, publicAxios }}>
      {children}
    </Provider>
  );
};

export { AxiosContext, AxiosProvider };
