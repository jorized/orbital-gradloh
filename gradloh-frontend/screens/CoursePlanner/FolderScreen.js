import React, { useContext, useEffect, useRef, useState } from 'react'
import { ActivityIndicator, Dimensions, FlatList, StyleSheet, Text, TouchableOpacity, View } from 'react-native'
import DrawerHeader from '../../components/Drawer/DrawerHeader'
import ThemeContext from '../../contexts/ThemeContext';
import { Animations } from '../../misc/consts';
import * as Animatable from 'react-native-animatable'
import { Entypo } from '@expo/vector-icons';
import { AxiosContext } from '../../contexts/AxiosContext';
import * as SecureStore from 'expo-secure-store';
import { EventRegister } from 'react-native-event-listeners';

export default function FolderScreen({ headerName, navigation }) {

    const publicAxios = useContext(AxiosContext);

    const userProfileDetails = SecureStore.getItem('userprofiledetails');
    const email = JSON.parse(userProfileDetails).email;

    const viewRef = useRef(null);
    const animation = Animations[15]

    const theme = useContext(ThemeContext);

    const [loading, setLoading] = useState(true);
    const [numOfModsInEachFolder, setNumOfModsInEachFolder] = useState({});
    const [currentSemester, setCurrentSemester] = useState(0);

    const fetchAndAnimate = () => {
        publicAxios.authAxios.get('/allfolderdetails', {
            params: { email: email }
        }).then(response => {
            const data = response.data;
            setCurrentSemester(data.currentSemester);
            setNumOfModsInEachFolder(data.numOfModsInEachFolder);
            setLoading(false);
        }).catch(error => {
            console.log(error);
        }).then(() => {
            const unsubscribe = navigation.addListener('focus', () => {
                viewRef.current.animate({ 0: { opacity: 0.5, }, 1: { opacity: 1 } });
              })
              return () => unsubscribe;
        })
    }

    useEffect(() => {
        fetchAndAnimate();
        const listener = EventRegister.addEventListener('updateScreens', (data) => {
            // Handle the event and update state as needed
            fetchAndAnimate(); // Re-fetch data or update state as needed
        });
    
        return () => {
            EventRegister.removeEventListener(listener);
        };


    }, [])

    const semesterMapping = [
        "Y1S1", "Y1S2", "Y2S1", "Y2S2", "Y3S1", "Y3S2", "Y4S1", "Y4S2"
    ];
    
    const ListItem = ({ item, index, animation, navigation, folderCount, currentSemester  }) => {
        return (
            <Animatable.View
            animation={animation}
            duration={1000}
            delay={index * 100}
            >
                <TouchableOpacity
                    style={[styles.listItem, { backgroundColor: theme.gradientColorBottomSheet }]}
                    activeOpacity={0.7}
                    onPress={() => navigation.push('FolderDetailsScreen', { headerName : semesterMapping[index], semIndex : index + 1})}>
                    <View style={styles.iconContainer}>
                        {currentSemester === index + 1 && (
                            <Text style={styles.currentLabel}>Current Sem</Text>
                        )}
                        <Entypo name="folder" size={94} color={theme.hamburgerColor} />
                        <Text style={[styles.iconLabel, { color: theme.color }]}>{folderCount}</Text>
                    </View>
                    <View style={styles.detailsContainer}>
                        <Text style={[styles.name, { color: theme.color }]}>{semesterMapping[index]}</Text>
                    </View>
                </TouchableOpacity>

        </Animatable.View>
        );
    }

    const renderItem = ({ item, index }) => (
        <ListItem item={item} index={index} animation={animation} navigation={navigation} folderCount={numOfModsInEachFolder[index + 1] || 0} currentSemester={currentSemester}/>)


    return (
        <View style={[styles.container, {backgroundColor: theme.backgroundColor}]}>
            <DrawerHeader headerName={headerName}/>
            {loading ? <ActivityIndicator size="small" color={theme.hamburgerColor} /> : 
            <Animatable.View
                ref={viewRef}
                easing={'ease-in-out'}
                duration={150}
                style={styles.container}>
                <FlatList
                data={Array(8).fill('')}
                keyExtractor={(_, i) => String(i)}
                numColumns={2}
                renderItem={renderItem}
                showsVerticalScrollIndicator={false}
                contentContainerStyle={{ paddingBottom: 100 }}
                />
            </Animatable.View>}
        </View>
    )
}
const styles = StyleSheet.create({
    container: {
        flex: 1,
    },
    name: {
        fontWeight: 'bold',
        fontSize: 16,
    },
    separator: {
        height: StyleSheet.hairlineWidth,
        backgroundColor: 'rgba(0, 0, 0, .08)',
    },
    listEmpty: {
        height: Dimensions.get('window').height,
        alignItems: 'center',
        justifyContent: 'center',
    },
    listItem: {
        height: 200,
        width: Dimensions.get('window').width / 2 - 16,
        backgroundColor: 'white',
        margin: 8,
        borderRadius: 10,
    },
    iconContainer: {
        height: 150,
        margin: 5,
        borderRadius: 10,
        justifyContent: "center",
        alignItems: "center",
        position: 'relative',
    },
    iconLabel: {
        position: 'absolute',
        fontSize: 24,
        fontWeight: 'bold',
        top: 70
    },
    detailsContainer: {
        paddingHorizontal: 16,
        paddingVertical: 5,
        flexDirection: 'row',
        justifyContent: 'space-between',
        alignItems: 'center',
        justifyContent: "center"
    },
    currentLabel: {
        position: 'absolute',
        top: 0,
        fontSize: 16,
        fontWeight: 'bold',
        color: 'red', 
    },
});
