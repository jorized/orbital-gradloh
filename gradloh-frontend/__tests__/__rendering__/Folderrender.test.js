import React from 'react';
import TestRenderer, { act } from 'react-test-renderer';
import { NavigationContainer, useNavigation, useRoute } from '@react-navigation/native';
import FolderScreen from '../../screens/CoursePlanner/FolderScreen';
import { AxiosContext } from '../../contexts/AxiosContext';
import ThemeContext from '../../contexts/ThemeContext';
import { LoadingContext } from '../../contexts/LoadingContext';

jest.useFakeTimers();

jest.mock('react-native-toast-message', () => {
  const React = require('react');
  const Toast = React.forwardRef((props, ref) => <div ref={ref} {...props} />);
  return {
    show: jest.fn(),
    hide: jest.fn(),
    default: Toast,
  };
});

jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: () => ({ goBack: jest.fn(), navigate: jest.fn(), dispatch: jest.fn() }),
  useRoute: () => ({
    params: {
      headerName: 'All Folders',  // Ensure this matches the expected pattern for your regex
      startTutorial: false
    }
  }),
}));

jest.mock('expo-secure-store', () => ({
    getItem: jest.fn().mockImplementation(() => JSON.stringify({ email: 'test@u.nus.edu' }))
}));

const mockAxiosContextValue = {
    authAxios: {
      get: jest.fn().mockResolvedValue({
        data: { numOfModsInEachFolder: { 1: 5, 2: 3, 3: 4, 4: 2, 5: 1, 6: 0, 7: 0, 8: 0 }, currentSemester: 1 }
      }),
    },
  };
  
const mockThemeContextValue = {
    backgroundColor: '#fff',
    reverseColor: '#000',
    hamburgerColor: '#ccc',
    gradientColorBottomSheet: '#eee',
    color: '#333',
  };
  
const mockLoadingContextValue = {
    setIsLoading: jest.fn()
  };

describe('FolderScreen', () => {
    it('renders correctly', async () => {
      let tree;

      const mockedParams = {
        route: {     
                params: {
                headerName: 'All Folders', 
                startTutorial: false
            } 
        },
      };

      await act(async () => {
        tree = TestRenderer.create(
          <LoadingContext.Provider value={mockLoadingContextValue}>
            <AxiosContext.Provider value={mockAxiosContextValue}>
              <ThemeContext.Provider value={mockThemeContextValue}>
                <NavigationContainer>
                  <FolderScreen 
                    headerName="All Folders" 
                    navigation={{
                        navigate: jest.fn(),
                        addListener: jest.fn().mockImplementation((event, callback) => {
                          callback();
                          //returning value for `navigationSubscription`
                          return {
                             remove: jest.fn()
                          }
                        }),
                      }}
                    {...mockedParams}/>
                </NavigationContainer>
              </ThemeContext.Provider>
            </AxiosContext.Provider>
          </LoadingContext.Provider>
        ).toJSON();
      });

      expect(tree).toMatchSnapshot();
    });
});
