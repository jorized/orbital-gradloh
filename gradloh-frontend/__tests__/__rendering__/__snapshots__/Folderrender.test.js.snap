// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FolderScreen renders correctly 1`] = `null`;
